internal/registry/registry.go

```diff
@@ -123,52 +119,94 @@ func (r *registry) getImageManifests(ctx context.Context, tag, platform string)
 		}
-		return r.getMultiPlatformManifests(ctx, &index, platform)
+		return r.findPlatformManifest(ctx, &index, tag, platform)
 	case strings.Contains(acceptImageManifestV1, mediaType):
 		manifest := imageManifestV1{}
 		if err = json.Unmarshal(b, &manifest); err != nil {
 			return nil, fmt.Errorf("error unmarshalling image manifest from %s: %w", url, err)
 		}
 		manifest.URL = url
-		return []*imageManifestV1{&manifest}, nil
+		return &manifest, nil
 	default:
 		return nil, fmt.Errorf("unknown mediaType %s from %s: %w", mediaType, url, err)
 	}
 }
 
-func (r *registry) getMultiPlatformManifests(ctx context.Context, index *imageIndexV1, platform string) ([]*imageManifestV1, error) {
-	var manifests []*imageManifestV1 //nolint:prealloc
+func (r *registry) findPlatformManifest(ctx context.Context, index *imageIndexV1, tag, platform string) (*imageManifestV1, error) {
+	platformToURL := map[string]string{} // duplicate keys are possible with os.version
+	platformToOSVersion := map[string]string{}
+	urlToMediaType := map[string]string{}
+
 	for _, ref := range index.Manifests {
 		p := pathutil.Join(ref.Platform.OS, ref.Platform.Architecture)
-		if p != platform {
-			continue
-		}
 		url := fmt.Sprintf("%s/manifests/%s", r.baseURL, ref.Digest)
-		manifest := imageManifestV1{}
-		if err := r.httpClient.GetJSON(ctx, url, ref.MediaType, &manifest); err != nil {
-			return nil, fmt.Errorf("error getting image ref for platform %v: %w", ref.Platform, err)
+
+		lastOSVersion := platformToOSVersion[p]
+		if ref.Platform.OSVersion >= lastOSVersion {
+			platformToURL[p] = url
+			urlToMediaType[url] = ref.MediaType
+			platformToOSVersion[p] = ref.Platform.OSVersion
 		}
-		manifest.URL = url
-		manifests = append(manifests, &manifest)
 	}
-	return manifests, nil
+
+	var err error
+	if platform, err = requireValidPlatform(tag, platform, platformToURL); err != nil {
+		return nil, err
+	}
+
+	url := platformToURL[platform]
+	mediaType := urlToMediaType[url]
+
+	manifest := imageManifestV1{}
+	if err := r.httpClient.GetJSON(ctx, url, mediaType, &manifest); err != nil {
+		return nil, fmt.Errorf("error getting image ref for platform %s: %w", platform, err)
+	}
+	manifest.URL = url
+	return &manifest, nil
 }
 
-func (r *registry) getImageConfigs(ctx context.Context, images []*imageManifestV1) ([]*imageConfigV1, error) {
-	var configs = make([]*imageConfigV1, len(images))
-	for i, image := range images {
-		if !strings.Contains(acceptImageConfigV1, image.Config.MediaType) {
-			return nil, fmt.Errorf("invalid config media type in image %v", image)
+func requireValidPlatform(tag, platform string, platforms map[string]string) (string, error) {
+	// create a list of the platforms we found
+	var s []string //nolint:prealloc
+	for p := range platforms {
+		s = append(s, p)
+	}
+	sort.Slice(s, func(i, j int) bool {
+		return s[i] < s[j]
+	})
+
+	if platform == "" {
+		switch len(platforms) {
+		case 0:
```

Comment:
I think we can return `return "", fmt.Errorf("tag %s has no platform information", tag)` here rather than check len(platforms) == 0 below.